File: OrbiterKEP.cpp
#include "OrbiterKEPConfig.h"

#include <string>
#include <iostream>

#include <keplerian_toolbox/planet/jpl_low_precision.h>
#include <keplerian_toolbox/planet/spice.h>
#include <keplerian_toolbox/util/spice_utils.h>
#include <pagmo/pagmo.h>

#include "src/param.h"
#include "src/problems/mga_transx.h"
#include "src/problems/mga_1dsm_transx.h"
#include "src/optimise.h"

#ifdef BUILD_MONGODB

#include "src/mongo/orbiter_db.h"

#endif

namespace orbiterkep {
  

} // namespaces

void run_problem(pagmo::problem::TransXSolution * solution, const pagmo::problem::transx_problem &single_obj, const pagmo::problem::transx_problem &multi_obj, int trials, int gen, double max_deltav, bool run_multi_obj) {
      int mf = 150;

      pagmo::decision_vector sol_mga;

      std::cout << "- Running single-objective optimisation";
      orbiterkep::optimiser op(single_obj, trials, gen, mf, 1);
      sol_mga = op.run_once(0, false, max_deltav);
      std::cout << " Done" << std::endl;

      if (run_multi_obj) {
        std::cout << "- Running multi-objective optimisation";
        orbiterkep::optimiser op_multi(multi_obj, trials, gen, mf, 1);
        op_multi.run_once(&sol_mga, true, max_deltav);
        std::cout << "Done" << std::endl;
      }

      single_obj.fill_solution(solution, sol_mga);
}

int main(int argc, char **argv) {
  try {
    orbiterkep::orbiterkep_db db;

    orbiterkep::parameters param = orbiterkep::parse_parameters(argc, argv);

    std::cout << std::fixed;
    std::cerr << std::fixed;

    if (param.n_mga > 0) {
      std::cout << "Executing MGA trials" << std::endl;

      pagmo::problem::mga_transx mga(param.planets, 
          param.dep_altitude, param.arr_altitude, param.circularize, 
          param.t0[0], param.t0[1], param.tof[0], param.tof[1], 
          param.vinf[0], param.vinf[1],
          param.add_dep_vinf, param.add_arr_vinf,
          false);

      pagmo::problem::mga_transx mga_multi(param.planets,
          param.dep_altitude, param.arr_altitude, param.circularize, 
          param.t0[0], param.t0[1], param.tof[0], param.tof[1], 
          param.vinf[0], param.vinf[1],
          param.add_dep_vinf, param.add_arr_vinf,
           true);

      pagmo::problem::TransXSolution solution;
      if (param.use_db) {
        mga.fill_solution(&solution, db.get_stored_solution(param, "MGA"));
      } else {
        run_problem(&solution, mga, mga_multi, param.n_mga, param.n_gen, param.max_deltaV, param.multi_obj);

        db.store_solution(param, solution, "MGA");
      }

      std::cout << solution << std::endl;

    }
    if (param.n_mga_1dsm > 0) {
      std::cout << "Executing MGA-1DSM trials" << std::endl;

      pagmo::problem::mga_1dsm_transx mga_1dsm(param.planets,
          param.dep_altitude, param.arr_altitude, param.circularize, 
          param.t0[0], param.t0[1], param.tof[0], param.tof[1], 
          param.vinf[0], param.vinf[1],
          param.add_dep_vinf, param.add_arr_vinf,
          false, true);


      pagmo::problem::mga_1dsm_transx mga_1dsm_multi(param.planets,
          param.dep_altitude, param.arr_altitude, param.circularize, 
          param.t0[0], param.t0[1], param.tof[0], param.tof[1], 
          param.vinf[0], param.vinf[1],
          param.add_dep_vinf, param.add_arr_vinf,
          true, true);


      pagmo::problem::TransXSolution solution;
      if (param.use_db) {
        mga_1dsm.fill_solution(&solution, db.get_stored_solution(param, "MGA-1DSM"));
      } else {
        run_problem(&solution, mga_1dsm, mga_1dsm_multi, param.n_mga_1dsm, param.n_gen, param.max_deltaV, param.multi_obj);

        db.store_solution(param, solution, "MGA-1DSM");
      }

      std::cout << solution << std::endl;
    }
  } catch (TCLAP::ArgException &e) {
    std::cerr << "error: " << e.error() << " for arg " << e.argId() << std::endl;
  }
  return 0;
}

File: python_legacy/OrbiterKEP.py
from PyGMO import archipelago, problem, island
from PyGMO.algorithm import jde,mbh,cs
from PyGMO.topology import ring, fully_connected
from PyKEP import epoch, epoch_from_iso_string
from PyKEP.planet import jpl_lp,spice
from PyKEP.trajopt import mga_1dsm
from PyKEP.util import load_spice_kernel
from problems.mga_1dsm_transx import mga_1dsm_transx
from problems.mga_transx import mga_transx 
from scipy.linalg import norm
import matplotlib.pylab as plt
import argparse
import logging
import sys
import os

from optimise import optimizer

parser.add_argument('--ship-max-deltaV', dest='max_deltav', help='your ship\'s Delta-V budget', default = 20000)

args = parser.parse_args()

for root, dirs, filenames in os.walk('spice'):
  for f in filenames:
    load_spice_kernel('./spice/' + f)

launch = [epoch_from_iso_string(x) for x in args.launch.split(',')]
tof = [float(x) for x in args.tof.split(',')]
vinf = [0.5, float(args.vinf_max)]

# We define an Earth-Venus-Earth problem (single-objective)
seq = [jpl_lp(name) for name in args.planets.split(',')]
# seq = [spice(name) for name in args.planets.split(',')]
# seq = [jpl_lp('earth'), jpl_lp('venus'), jpl_lp('venus'), jpl_lp('earth'), jpl_lp('jupiter'), jpl_lp('saturn')]

# enc_times = [50736.36319, 50929.5, 51353.5, 51408.14444, 51908.5, 53187.11667]
enc_times = []

runMGA = True
runMGA_1DSM = True 

sol_mga = None
t = None

nTrialMGA = int(args.n_mga)
nTrialMGA1DSM = int(args.n_mga_1dsm)

multi_obj = args.multi_obj

sol_mga_1dsm = None
try:
  if runMGA:
    mga = mga_transx(seq=seq, 
                     dep_altitude = args.dep_altitude, 
                     arr_altitude = args.arr_altitude, 
                     circularize = args.circularize,
                     t0 = launch, tof = tof, 
                     add_vinf_dep = args.add_dep_vinf, add_vinf_arr = args.add_arr_vinf,
                     multi_objective = False)
    mga_multi = None
    if multi_obj:
      mga_multi = mga_transx(seq=seq, 
                             dep_altitude = args.dep_altitude, 
                             arr_altitude = args.arr_altitude, 
                             circularize = args.circularize,
                             t0 = launch, tof = tof, 
                             add_vinf_dep = args.add_dep_vinf, add_vinf_arr = args.add_arr_vinf,
                             multi_objective = multi_obj)
    solutions = []
    for i in range(nTrialMGA):
      print("MGA problem - Iteration %d" % (i + 1))
      # mga.avoid = solutions
      op = optimizer(mga, nTrial = 5, gen = 1000, mf = 100, mr = 1)
      sol_mga = op.run_once()
      if multi_obj:
        print("MGA problem - Multi-Objective phase - Iteration %d" % (i + 1))
        op = optimizer(mga_multi, 1, gen = 4000, mf = 100, mr = 10)
        sol_mga_multi = op.run_once(sol_mga, True)

      if sol_mga != None:
        solutions.append(sol_mga)

    if (len(solutions) > 0):
      sol_mga = min(solutions, key=lambda x: mga.calc_objective(x))

  if runMGA_1DSM:
    mga_1dsm = mga_1dsm_transx(seq=seq, 
                               dep_altitude = args.dep_altitude, 
                               arr_altitude = args.arr_altitude, 
                               circularize = args.circularize,
                               t0=launch, tof = tof, vinf=vinf, 
                               add_vinf_dep = args.add_dep_vinf, add_vinf_arr = args.add_arr_vinf, 
                               multi_objective = False)
    mga_1dsm_multi = None

    if multi_obj:
      mga_1dsm_multi = mga_1dsm_transx(seq=seq, 
                                       dep_altitude = args.dep_altitude, 
                                       arr_altitude = args.arr_altitude, 
                                       circularize = args.circularize,
                                       t0=launch, tof = tof, vinf = vinf, 
                                       add_vinf_dep = args.add_dep_vinf, add_vinf_arr = args.add_arr_vinf, 
                                       multi_objective = multi_obj)
    solutions = []
    # if sol_mga != None:
    #   t0 = sol_mga[0]
    #   T = sol_mga[1] / 365.25
    #   launch = [epoch(t0 - 30), epoch(t0 + 30)]
    #   tof = [0.1, T * 1.25]
    for i in range(nTrialMGA1DSM):
      # mga_1dsm.avoid = solutions
      op = optimizer(mga_1dsm, nTrial=5, gen=1000, mf=1000, mr = 2)
      sol_mga_1dsm = op.run_once()
      if sol_mga_1dsm != None:
        solutions.append(sol_mga_1dsm)
      
      if multi_obj:
        op = optimizer(mga_1dsm_multi, nTrial=1, gen=4000, mf = 100, mr = 10)
        sol_mga_multi = op.run_once(sol_mga_1dsm, True)

    if (len(solutions) > 0):
      sol_mga_1dsm = min(solutions, key=lambda x: mga_1dsm.calc_objective(x))
except Exception, e:
  logging.error(e, exc_info = True)
finally:
  if sol_mga != None:
    mga.pretty(sol_mga)
  if sol_mga_1dsm != None:
    mga_1dsm.pretty(sol_mga_1dsm)

